### PR TITLE
bem-xjst: short syntax for simple modes

### LINKED ISSUE #444: Short syntax for simple modes
Before:
```js
block('a').tag()('span')
```

After:
```js
block('a')({ tag: 'span' })
```

Before:
```js
block('a')(
    tag()('span'),
    addMix()({ block: 'x' }),
    mods()({ mod: 'val' }),
    content()(function() {
        return ['Vobla: ', this.ctx.name];
    })
});
```

After:
```js
block('a')({
    tag: 'span',
    addMix: { block: 'x' },
    mods: { mod: 'val' },
    content() {
        return ['Vobla: ', this.ctx.name];
    }
});
```

Complex cases gonna be like this.

Before:
```js
block('a')(
    tag()('a'),
    match(function() { return this.bla }).tag()('span')
);
```

After:
```js
block('a')(
    { tag: 'a' },
    match(function() { return this.bla })({
        tag: 'span'
    })
);
```

Some cases gonna be only in old syntax:
```js
block('a').tag()(
    'span',
    match(function() { })('a')
)
```

### PR DESCRIPTION
**Pull request by miripiruni**
Fix for #444

### Changes proposed in this pull request
- It’s a minor. No breaking major changes. I can port the changes to v7.x if needed. 
- New shortcuts visualy separates output generators (content, attrs, mix, and etc) and conditions (match, mod, elem, etc).
- New way of thinking in documentation fabula. Learn only new object-like syntax for output generators and use conditions in chain or composition.
- Try it in online demo: [soon]

### Some nuances

- In shortcut syntax `default` stands for `def()`. Because of [inner naming](https://github.com/bem/bem-xjst/blob/c897cc03eca7f7d3e4e5b08f0a8618288b01575c/lib/bemxjst/entity.js#L51). I created shortcut `def` for `default`, but `default` key is already in use.
- Earlier in documentation and presentations I used notation `modeName` + `()` to mark modes. Example: `attrs()`. Now we should stop doing this, because the main way of writing becomes object-like shortcut.
- `match` in shortcut is [treated](https://github.com/bem/bem-xjst/pull/497/files#diff-b23aaab454fd9970aa06b9d1936facb2R106) as user defined mode not as subpredicate. 

### Checklist

 - [x] Documentation: ru
 - [x] Documentation: en
 - [x] Integration test ([on turbo](https://nda.ya.ru/3Twq6B))
 - [x] Tests added

### Bench

`templates.apply()` — no diff. There is no runtime effect. Only compile staff.

`templates.compile()` — no diff with old templates.

### CODE REVIEW

**Comment by zxqfox** [comment]
> New way of thinking in documentation fabula. Learn only new object-like syntax for output generators and use conditions in chain or composition.

This is freaking awesome.

File: docs/en/3-api.md [hunk 1]
```diff
@@ -360,11 +362,13 @@ var bemxjst = require('bem-xjst');
...
 
 var templates = bemhtml.compile(function() {
-  block('b').content()('yay');
+  block('b')({ content: 'yay' });
 
-  block('mods-changes').def()(function() {
-    this.ctx.mods.one = 2;
-    return applyNext();
+  block('mods-changes')({
+      default: function() {
```
Comment (zxqfox) [inline docs/en/3-api.md]: Maybe like that?
```js
default(_, ctx) {
  ctx.mods.one = 2; // Why you modifying this.ctx.mods but no this.mods?
  return applyNext(); // Can we make an alias here to call as this.__base() ?
}
```

File: docs/en/3-api.md [hunk 2]
```diff
@@ -407,10 +411,12 @@ You can use option `production` to render whole BEMJSON even if one template con
...
 var template = bemxjst.compile(function() {
-  block('b1').attrs()(function() {
-    var attrs = applyNext();
-    attrs.undef.foo = 'bar';
-    return attrs;
+  block('b1')({
+      attrs: function() {
+          var attrs = applyNext();
+          attrs.undef.foo = 'bar';
+          return attrs;
```
Comment (zxqfox) [inline docs/en/3-api.md]: Almost like that:
```js
attrs({ __base }) {
  return { ...__base(), undef: { foo: 2 } };
}
```

File: docs/en/3-api.md [hunk 3]
```diff
@@ -511,11 +519,13 @@ In templates body the module will be acessible as `this.require('moment')`.
 You can use the template in any browser or in `Node.js`:
 
 ```js
-block('post').elem('data').content()(function() {
-    var moment = this.require('moment');
-
-    return moment(this.ctx.date) // Time in ms from server
-        .format('YYYY-MM-DD HH:mm:ss');
+block('post').elem('data')({
+    content: function() {
```
Comment (zxqfox) [inline docs/en/3-api.md]: same: `content() {`

File: docs/ru/7-runtime.md [hunk 4]
```diff
@@ -198,11 +203,13 @@ applyCtx(bemjson, newctx)
 Шаблон:
 
 ```js
-block('header').def()(function() {
-    return applyCtx(this.extend(this.ctx, {
-        block: 'layout',
-        mix: [{ block: 'header' }].concat(this.ctx.mix || [])
-    }));
+block('header')({
+    default: function() {
```
Comment (zxqfox) [inline docs/ru/7-runtime.md]: It won't work in few IEs, right? Cuz of it's a protected reserved word: https://msdn.microsoft.com/en-us/library/ttyab5c8.aspx

Ie8 for sure, not sure about ie9-11.

File: .jshintrc [hunk 5]
```diff
@@ -99,6 +87,41 @@
     "globals"       : {
         "module": true,
         "test": true,
-        "BEMXJSTError": true
+        "BEMXJSTError": true,
+        // Subpredicates:
```
[the remark this answers is not in the document]
Comment (miripiruni) [reply .jshintrc]: @zxqfox because jshint fields `predef` and `globals` are the same for this case.

File: docs/ru/7-runtime.md (hunk 4 above)
Comment (miripiruni) [reply docs/ru/7-runtime.md]: @zxqfox anyway it work if you use quotes ;)

**Comment by zxqfox** [comment]
Awesome new everyone!

**Pull request merged**